flash/image/classification/integrations/baal/loop.py

```diff
@@ -0,0 +1,162 @@
+            # reload the weights to retrain from scratch with the new labelled data.
+            self.trainer.lightning_module.load_state_dict(self._model_state_dict)
+        self.progress.increment_completed()
+        return super().on_advance_end()
+
+    def on_run_end(self):
+        self.trainer.lightning_module.predict_step = self.trainer.lightning_module._predict_step
+        return super().on_run_end()
+
+    def on_save_checkpoint(self) -> Dict:
+        return {"datamodule_state_dict": self.trainer.datamodule.state_dict()}
+
+    def on_load_checkpoint(self, state_dict) -> None:
+        self.trainer.datamodule.load_state_dict(state_dict.pop("datamodule_state_dict"))
+
+    ############## UTILS ##############
+
+    def __getattr__(self, key):
+        if key not in self.__dict__:
+            return getattr(self.fit_loop, key)
+        return self.__dict__[key]
+
+    def _reset_fitting(self):
+        self.trainer.state.fn = TrainerFn.FITTING
+        self.trainer.state.status = TrainerStatus.RUNNING
+        self.trainer.training = True
+        self.trainer.lightning_module.on_train_dataloader()
+
+    def _reset_predicting(self):
+        self.trainer.state.fn = TrainerFn.PREDICTING
+        self.trainer.state.status = TrainerStatus.RUNNING
+        self.trainer.predicting = True
+        self.trainer.lightning_module.on_predict_dataloader()
+        self._enable_mc_dropout()
+
+    def _reset_dataloader_for_stage(self, running_state: RunningStage):
+        dataloader_name = f"{running_state.value}_dataloader"
+        setattr(
+            self.trainer.lightning_module,
+            dataloader_name,
+            _PatchDataLoader(getattr(self.trainer.datamodule, dataloader_name)(), running_state),
+        )
+        setattr(self.trainer, dataloader_name, None)
+        getattr(self.trainer, f"reset_{dataloader_name}")(self.trainer.lightning_module)
+
+    def _enable_mc_dropout(self):
+        # prevent the model to put into val model - hack
+        self.trainer.lightning_module.on_predict_model_eval = self._do_nothing
+        for _, module in self.trainer.lightning_module.named_modules():
+            if isinstance(module, _BatchNorm):
+                module.eval()
+        # save the predict_step and replace it within a mc_inference function
+        if getattr(self.trainer.lightning_module, "_predict_step", None) is None:
+            self.trainer.lightning_module._predict_step = self.trainer.lightning_module.predict_step
+        self.trainer.lightning_module.predict_step = partial(
+            self._mc_inference,
+            predict_step_fn=self.trainer.lightning_module._predict_step,
+            inference_iteration=self.inference_iteration,
+        )
+
+    def _do_nothing(self, *args, **kwargs):
+        pass
+
+    @staticmethod
+    def _mc_inference(*args, predict_step_fn: Callable = None, inference_iteration: int = None, **kwargs):
+        out = []
+        for _ in range(inference_iteration):
```

Comment:
BaaL expects a shape [num_samples, num_classes, num_iterations]

Reply:
Good catch !

Reply:
I tried to use it this way, it seems to break